pkg/rbd/rbd_util.go

```diff
@@ -205,10 +185,491 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+		ClusterID:       rbdVol.ClusterID,
+		ObjectUUID:      imageUUID,
+	}
+	rbdVol.VolID, err = vi.ComposeCsiID()
+	if err != nil {
+		return false, err
+	}
+
+	klog.V(4).Infof("Found existng volume (%s) with image name (%s) for request (%s)",
+		rbdVol.VolID, rbdVol.RbdImageName, rbdVol.RequestName)
+
+	return true, nil
+}
+
+/*
+unreserveSnap and unreserveVol remove omaps associated with the snapshot and the image name,
+and also remove the corresponding request name key in the snaps or volumes omaps respectively.
+
+This is performed within the request name lock, to ensure that requests with the same name do not
+manipulate the omap entries concurrently.
+*/
+func unreserveSnap(rbdSnap *rbdSnapshot, credentials map[string]string) error {
+	key, err := getKey(rbdSnap.AdminID, credentials)
+	if err != nil {
+		return err
+	}
+
+	// delete snap image omap (first, inverse of create order)
+	snapUUID := strings.TrimPrefix(rbdSnap.RbdSnapName, rbdSnapNamePrefix)
+	err = util.RemoveObject(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool, rbdSnapOMapPrefix+snapUUID)
+	if err != nil {
+		if _, ok := err.(util.ErrObjectNotFound); !ok {
+			klog.V(4).Infof("failed removing oMap %s (%s)", rbdSnapOMapPrefix+snapUUID, err)
+			return err
+		}
+	}
+
+	// delete the request name omap key (last, inverse of create order)
+	err = util.RemoveOMapKey(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+		csiSnapsDirectory, csiSnapNameKeyPrefix+rbdSnap.RequestName)
+	if err != nil {
+		klog.V(4).Infof("failed removing oMap key %s (%s)", csiSnapNameKeyPrefix+rbdSnap.RequestName, err)
+		return err
+	}
+
+	return nil
+}
+
+func unreserveVol(rbdVol *rbdVolume, credentials map[string]string) error {
+	key, err := getKey(rbdVol.AdminID, credentials)
+	if err != nil {
+		return err
+	}
+
+	// delete image omap (first, inverse of create order)
+	imageUUID := strings.TrimPrefix(rbdVol.RbdImageName, rbdImgNamePrefix)
+	err = util.RemoveObject(rbdVol.Monitors, rbdVol.AdminID, key, rbdVol.Pool, rbdImageOMapPrefix+imageUUID)
+	if err != nil {
+		if _, ok := err.(util.ErrObjectNotFound); !ok {
+			klog.V(4).Infof("failed removing oMap %s (%s)", rbdImageOMapPrefix+imageUUID, err)
+			return err
+		}
+	}
+
+	// delete the request name omap key (last, inverse of create order)
+	err = util.RemoveOMapKey(rbdVol.Monitors, rbdVol.AdminID, key, rbdVol.Pool,
+		csiVolsDirectory, csiVolNameKeyPrefix+rbdVol.RequestName)
+	if err != nil {
+		klog.V(4).Infof("failed removing oMap key %s (%s)", csiVolNameKeyPrefix+rbdVol.RequestName, err)
+		return err
+	}
+
+	return nil
+}
+
+// reserveOMapName creates an omap with passed in oMapNamePrefix and a generated <uuid>.
+// It ensures generated omap name does not already exist and if conflicts are detected, a set
+// number of retires with newer uuids are attempted before returning an error
+func reserveOMapName(monitors, adminID, key, poolName, oMapNamePrefix string) (string, error) {
+	var iterUUID string
+
+	maxAttempts := 5
```

Comment:
No reason, if there is a UUID clash that is not resolved in 5 retries, it may never be :) . Further, the CO would retry anyway to get another 5 tries to get a new UUID.